menu.lisp

```diff
@@ -188,51 +197,45 @@ Returns the selected element in TABLE or nil if aborted. "
   (check-type table list)
   (check-type prompt (or null string))
   (check-type initial-selection integer)
-  (let* ((menu-options (mapcar #'menu-element-name table))
-         (menu-require-scrolling (and *menu-maximum-height*
-                                       (> (length menu-options)
-                                          *menu-maximum-height*)))
-         (menu (make-menu-state
-                :table table
-                :prompt prompt
-                :view-start (if menu-require-scrolling initial-selection 0)
-                :view-end (if menu-require-scrolling
-                              (+ initial-selection *menu-maximum-height*)
-                              (length menu-options))
-                :selected initial-selection))
-         (*record-last-msg-override* t)
-         (*suppress-echo-timeout* t))
-    (bound-check-menu menu)
-    (catch :menu-quit
-      (unwind-protect
-           (with-focus (screen-key-window screen)
-             (loop
-                (let ((strings (subseq menu-options
+
+  (when table
+    (let* ((*record-last-msg-override* t)
+           (*suppress-echo-timeout* t)
+           (menu (make-menu-state
+                  :unfiltered-table table
+                  :table table
+                  :prompt prompt
+                  :view-start 0
+                  :view-end 0
+                  :selected 0))
+           (keymap (if extra-keymap
+                       (list extra-keymap *menu-map*)
+                       (list *menu-map*))))
+      (bound-check-menu menu)
+      (catch :menu-quit
+        (unwind-protect
+             (with-focus (screen-key-window screen)
+               (loop
+                  (let* ((prompt-row (format nil "~A ~A"
```

Comment:
A better format-string would be `"~@[~A ~]~A"`, because `PROMPT` can be `NIL` and if so, it shouldn't be printed.